src/lib/functions/runtimes/js/index.ts

```diff
@@ -3,45 +3,57 @@ import { dirname } from 'path'
 import { pathToFileURL } from 'url'
 import { Worker } from 'worker_threads'
 
-import lambdaLocal from 'lambda-local'
+import lambdaLocal, { type LambdaEvent } from 'lambda-local'
 
+import type { BuildFunction, GetBuildFunction, InvokeFunction, OnDirectoryScanFunction } from '../index.js'
 import { BLOBS_CONTEXT_VARIABLE } from '../../../blobs/blobs.js'
 import type NetlifyFunction from '../../netlify-function.js'
 
-import detectNetlifyLambdaBuilder from './builders/netlify-lambda.js'
-import detectZisiBuilder, { parseFunctionForMetadata } from './builders/zisi.js'
+import detectNetlifyLambdaBuilder, {
+  type NetlifyLambdaBuildResult,
+  type NetlifyLambdaBuilder,
+} from './builders/netlify-lambda.js'
+import detectZisiBuilder, { getFunctionMetadata, ZisiBuildResult } from './builders/zisi.js'
 import { SECONDS_TO_MILLISECONDS } from './constants.js'
-import { $TSFixMe } from '../../../../commands/types.js'
+import type { WorkerMessage } from './worker.js'
 
 export const name = 'js'
 
-// @ts-expect-error TS(7034) FIXME: Variable 'netlifyLambdaDetectorCache' implicitly h... Remove this comment to see the full error message
-let netlifyLambdaDetectorCache
+export type JsBuildResult = ZisiBuildResult | NetlifyLambdaBuildResult
+
+// TODO(serhalp) Unify these. This is bonkers that the two underlying invocation mechanisms are encapsulated but we
+// return slightly different shapes for them.
+export type JsInvokeFunctionResult = WorkerMessage | LambdaEvent
+
+let netlifyLambdaDetectorCache: undefined | NetlifyLambdaBuilder
 
 lambdaLocal.getLogger().level = 'alert'
 
 // The netlify-lambda builder can't be enabled or disabled on a per-function
 // basis and its detection mechanism is also quite expensive, so we detect
 // it once and cache the result.
-const detectNetlifyLambdaWithCache = () => {
-  // @ts-expect-error TS(7005) FIXME: Variable 'netlifyLambdaDetectorCache' implicitly h... Remove this comment to see the full error message
+const detectNetlifyLambdaWithCache = async () => {
   if (netlifyLambdaDetectorCache === undefined) {
-    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
+    netlifyLambdaDetectorCache = await detectNetlifyLambdaBuilder()
```

Comment:
this wasn't a sort of debounce mechanism or anything, it was just a plain result cache, so we don't need to hold on to the promise